### PR TITLE
[Dataset] Change finalize signature

### PR DESCRIPTION
**Pull request by zhoonit**
- [Dataset] Change finalize signature

```
This patch changes finalize signature to contain user_data,
user_data will be needed to replicate the current worker(with diffrent
arguments) in case of
MT_safty is not provided by the producer. Please note that this does not
change any current behavior but the preparation for the future
enhancements.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/dataset/random_data_producers.h [hunk 1]
```diff
@@ -82,7 +82,8 @@ class RandomDataOneHotProducer final : public DataProducer {
    */
```
Comment (lhs8928) [inline nntrainer/dataset/random_data_producers.h:82]: Please add arg user_data

File: nntrainer/dataset/raw_file_data_producer.h [hunk 2]
```diff
@@ -93,7 +93,8 @@ class RawFileDataProducer final : public DataProducer {
    */
```
Comment (lhs8928) [inline nntrainer/dataset/raw_file_data_producer.h:93]: same here

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/dataset/random_data_producers.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/dataset/random_data_producers.h:82]: Done

File: nntrainer/dataset/raw_file_data_producer.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/dataset/raw_file_data_producer.h:93]: Done

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**